### ISSUE TITLE
Geron Chapter 4

### ISSUE THREAD

**Issue by sccrosby** [sccrosby/python_data_explorations#2]
1. Stochastic Gradient Descent (GD) or mini batch GD are good options when you have millions of features
2. If features have varying scales the optimization space can have a very flat "gutter". It is good to scale variables first
3. Logistic Regressions are convex, and have no local minima
4. Local minima can be a problem for GD algorithms, other options like simulated annealing can be better in those cases
5. If validation error increases at every epoch during training the model is overfitting the training set, reduce the model parameters or add regularization to reduce this
6. ** Yes, It is a good idea to stop (early stopping) mini-batch gradient descent immediately when validation error goes up.
7. Stochastic GD will reach the vicinity of the solution fastest, but will bounce around and not converge. Batch GD will converge but is slow. Using a decrease learning rate (LR) or a LR schedule with stochastic GD can make it converge.
8. A gap between the training error and validation error means overfitting. To improve this reduce the number of features, add regularization, and add instances (data).
9. If training and validation errors are similar but both high the model suffers from high variance. You should reduce your regularization constraint
10. a. Ridge Regressions adds L2 regularization and will reduce overfitting. b. Lasso adds L1 regularization and will essentially remove less important features, however Lasso can converge poorly. Elastic Net adds a combo of L1 and L2, allowing for converging while setting less important features to near zero.
11. Logistic Classifiers can do not allow for multiple classifications, so you need two
